internal/component/azure/servicebus/receiver.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servicebus
+
+import (
+	"context"
+	"fmt"
+
+	azservicebus "github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
+	"go.uber.org/multierr"
+)
+
+type Receiver interface {
+	ReceiveMessages(ctx context.Context, maxMessages int, options *azservicebus.ReceiveMessagesOptions) ([]*azservicebus.ReceivedMessage, error)
+	CompleteMessage(ctx context.Context, m *azservicebus.ReceivedMessage, opts *azservicebus.CompleteMessageOptions) error
+	AbandonMessage(ctx context.Context, m *azservicebus.ReceivedMessage, opts *azservicebus.AbandonMessageOptions) error
+	Close(ctx context.Context) error
+}
+
+var (
+	_ Receiver = (*SessionReceiver)(nil)
+	_ Receiver = (*MessageReceiver)(nil)
+)
+
+func NewSessionReceiver(r *azservicebus.SessionReceiver) *SessionReceiver {
+	return &SessionReceiver{SessionReceiver: r}
+}
+
+type SessionReceiver struct {
+	*azservicebus.SessionReceiver
+}
+
+func (s *SessionReceiver) RenewSessionLocks(ctx context.Context) error {
+	if s == nil {
+		return nil
+	}
+
+	return s.RenewSessionLock(ctx, nil)
+}
+
+func NewMessageReceiver(r *azservicebus.Receiver) *MessageReceiver {
+	return &MessageReceiver{Receiver: r}
+}
+
+type MessageReceiver struct {
+	*azservicebus.Receiver
+}
+
+func (m *MessageReceiver) RenewMessageLocks(ctx context.Context, msgs []*azservicebus.ReceivedMessage) error {
+	if m == nil {
+		return nil
+	}
+
+	var errs []error
+	for _, msg := range msgs {
```

Comment:
I like this, only question now is: how many messages do you think could be refreshed in each iteration? Should there be a limit on the number of parallel goroutines we span, if you think there can be a lot of them?

Reply:
So in theory the max number of messages to be refreshed each iteration is limited by the `MaxActiveMessages` property which defaults to 1000. IMO It would be problematic to introduce a new conflicting constraint e.g. we can accept 1000 messages but only renew 250 as this would lead to cases where locks expire and multiple daprd instances process the same message. With this in mind maybe reducing the default max active messages to a lower number such as 250 would make sense. It would be unlikely people need to have over 250 inflight messages at any one time and if they did then they could up the limit accepting the consequences of increased resource consumption. Wdyt?